Refine classification section (#73)
carpentries-incubator/Natural-language-processing: episodes/03-transformers.md

```python

classifier = pipeline(task="text-classification", model="SamLowe/roberta-base-go_emotions", top_k=3)

sentences = ["I am not having a great day.", "This is a lovely and innocent sentence.", "Maria loves Groningen."]
model_outputs = classifier(sentences)

pretty_print_outputs(sentences, model_outputs)

=====	 I am not having a great day.
{'label': 'disappointment', 'score': 0.5044488906860352}
{'label': 'sadness', 'score': 0.3469431400299072}
{'label': 'annoyance', 'score': 0.08485794812440872}

=====	 This is a lovely and innocent sentence.
{'label': 'admiration', 'score': 0.7667419910430908}
{'label': 'approval', 'score': 0.4139412045478821}
{'label': 'love', 'score': 0.1130351722240448}

=====	 Maria loves Groningen.
{'label': 'love', 'score': 0.9160982370376587}
{'label': 'neutral', 'score': 0.07024713605642319}

> {'label': 'approval', 'score': 0.02595525234937668}

Reviewer: Just a suggestion: maybe these scores can be rounded to the third decimal place.

Author: I want to just log the exact output here, which contains all the decimals.

Changing the code to cut would be an option, but it would mean introducing extra functionality that is unrelated and hence will raise further questions.

```

For each sentence, this code outputs a list with the `top-k` (`k=3`) emotions to which the model has assigned the highest scores.
Because it is trained on a multi-label dataset, multiple labels can have a high score.

::: callout

In the process of fine-tuning, a generic language model like BERT is adapted to a specific task like text classification.
This is faster than training a whole model, because we only need to train the _classifier_ layer, a small neural network, that can learn to choose between the classes (labels) for your custom classification problem.

> This classifier is a one-layer neural layer that assigns scores over a set of labels. It uses the input features provided by BERT, which already _encodes_ the meaning of the entire sequence in its hidden states. You can learn more about fine-tuning BERT-like models in [this HuggingFace tutorial](https://huggingface.co/docs/transformers/v4.57.1/en/training#fine-tuning).

Reviewer: ```suggestion
This classifier is a single-layer neural network that assigns scores over a set of labels. It uses the input features provided by BERT, which already _encodes_ the meaning of the entire sequence in its hidden states. You can learn more about fine-tuning BERT-like models in [this HuggingFace tutorial](https://huggingface.co/docs/transformers/v4.57.1/en/training#fine-tuning).
```

Several metrics are commonly used to evaluate NLP models, each offering different insights into performance. Here we will describe the 4 most used in suervised learning classifiers

> Let's give the toy example of classifying sentiments conveyed in texts.

Reviewer: ```suggestion
Let's look at the toy example of classifying sentiments conveyed in texts.
```

>  In this task, there are three classes: `positive`, `neutral`, and `negative`. We have a model that return the label `positive`,`neutral` or `dog` given an input text. If we want to evaluate how good is our model at recognizing `positive`s, there are 4 possibilities when comparing with the ground truth (your labelled data): 

Reviewer: ```suggestion
 In this task, there are three classes: `positive`, `neutral`, and `negative`. We have a model that returns the label `positive`,`neutral` or `dog` given an input text. If we want to evaluate how good our model is at recognizing `positive`s, there are 4 possibilities when comparing with the ground truth (your labelled data): 
```

Author: Haha, I have additionally spotted a hidden `dog` leftover here.

We can see that the model predicts correctly the 4 examples we gave. This is unsurprising as they are incredibly obvious examples. We can also print the results and inspect them because they are only 4 instances, but it is clearly not a scalable approach. 

::: callout

> Note that many models will provide a score, with their predictions. It is very tempting to interpret these scores as a proxy to "how certain is the model of prediction X". However, this score is only a relative measure with respect to the training data, and it does not always translate well to unseen data.

Reviewer: ```suggestion
Note that many models assign scores to their predictions. It is very tempting to interpret these scores as a proxy for "how certain is the model of prediction X". However, this score is only a relative measure with respect to the training data, and it does not always translate well to unseen data.
```
